src/test/java/GmailServletTest.java

```diff
@@ -0,0 +1,114 @@
+ */
+@RunWith(JUnit4.class)
+public final class GmailServletTest {
+  private GmailClient gmailClient;
+  private GmailServlet servlet;
+
+  private static final boolean AUTHENTICATION_VERIFIED = true;
+  private static final String ID_TOKEN_KEY = "idToken";
+  private static final String ID_TOKEN_VALUE = "sampleId";
+  private static final String ACCESS_TOKEN_KEY = "accessToken";
+  private static final String ACCESS_TOKEN_VALUE = "sampleAccessToken";
+  private static final Cookie sampleIdTokenCookie = new Cookie(ID_TOKEN_KEY, ID_TOKEN_VALUE);
+  private static final Cookie sampleAccessTokenCookie =
+      new Cookie(ACCESS_TOKEN_KEY, ACCESS_TOKEN_VALUE);
+  private static final Cookie[] validCookies =
+      new Cookie[] {sampleIdTokenCookie, sampleAccessTokenCookie};
+
+  private HttpServletRequest request;
+  private HttpServletResponse response;
+  private StringWriter stringWriter;
+  private PrintWriter printWriter;
+
+  private static final String MESSAGE_ID_ONE = "messageIdOne";
+  private static final String MESSAGE_ID_TWO = "messageIdTwo";
+  private static final String MESSAGE_ID_THREE = "messageIdThree";
+  private static final List<Message> NO_MESSAGES = ImmutableList.of();
+  private static final List<Message> THREE_MESSAGES =
+      ImmutableList.of(
+          new Message().setId(MESSAGE_ID_ONE),
+          new Message().setId(MESSAGE_ID_TWO),
+          new Message().setId(MESSAGE_ID_THREE));
+  private static final String THREE_MESSAGES_JSON =
+      String.format(
+          "[{\"id\":\"%s\"},{\"id\":\"%s\"},{\"id\":\"%s\"}]",
+          MESSAGE_ID_ONE, MESSAGE_ID_TWO, MESSAGE_ID_THREE);
+  private static final String NO_MESSAGES_JSON = "[]";
+
+  @Before
+  public void setUp() throws IOException, GeneralSecurityException {
+    AuthenticationVerifier authenticationVerifier = Mockito.mock(AuthenticationVerifier.class);
+    GmailClientFactory gmailClientFactory = Mockito.mock(GmailClientFactory.class);
+    gmailClient = Mockito.mock(GmailClient.class);
+    servlet = new GmailServlet(authenticationVerifier, gmailClientFactory);
+
+    Mockito.when(gmailClientFactory.getGmailClient(Mockito.any())).thenReturn(gmailClient);
+    // Authentication will always pass
+    Mockito.when(authenticationVerifier.verifyUserToken(Mockito.anyString()))
+        .thenReturn(AUTHENTICATION_VERIFIED);
+
+    request = Mockito.mock(HttpServletRequest.class);
+    response = Mockito.mock(HttpServletResponse.class);
+    Mockito.when(request.getCookies()).thenReturn(validCookies);
+
+    // Writer used in get/post requests to capture HTTP response values
+    stringWriter = new StringWriter();
+    printWriter = new PrintWriter(stringWriter);
+    Mockito.when(response.getWriter()).thenReturn(printWriter);
+  }
+
+  @Test
+  public void noMessagesPresent() throws IOException, ServletException {
+    Mockito.when(gmailClient.listUserMessages(Mockito.anyString())).thenReturn(NO_MESSAGES);
```

Comment:
Can you restrict the mock to return the result only when the correct query is provided? That way, you'll be testing more parts of your code. Same thing below.

Reply:
Agreed - fixed! Thanks for catching this